valheim-updater

```diff
@@ -5,39 +5,81 @@ CONFIG_DIRECTORY_PERMISSIONS=${CONFIG_DIRECTORY_PERMISSIONS:-755}
+STEAMCMD_ARGS=${STEAMCMD_ARGS}
+FORCE_VALIDATE=${FORCE_VALIDATE:-false}
+VALHEIM_PLUS_ENABLED=${VALHEIM_PLUS_ENABLED:-true}
 just_started=true
 next_update=$(date +%s)
 
 cd /opt/steamcmd
 main() {
     while :; do
         update_permissions
-        update
+        update_valheim
+        if [ $VALHEIM_PLUS_ENABLED = true ]; then
+            update_valheim_plus
+        fi
         next_update=$(($(date +%s)+$UPDATE_INTERVAL))
         while [ $(date +%s) -lt $next_update ]; do
             sleep 5
         done
     done
 }
 
-update() {
-    local logfile="$(mktemp)"
-    echo "Updating/Validating Valheim Server"
-    ./steamcmd.sh +login anonymous +force_install_dir /opt/valheim_dl +app_update 896660 $STEAMCMD_ARGS +quit
-    rsync -a --itemize-changes --delete --exclude server_exit.drp --exclude steamapps /opt/valheim_dl/ /opt/valheim | tee "$logfile"
-    grep '^[*>]' "$logfile" > /dev/null 2>&1
-    if [ $? -eq 0 ]; then
-        echo "Valheim Server was updated - restarting"
-        supervisorctl restart valheim-server
+install_valheim_plus() {
+    echo "Installing/Updating Valheim Plus"
+    wget -q -O /opt/valheimplus/UnixServer.zip https://github.com/nxPublic/ValheimPlus/releases/latest/download/UnixServer.zip && \
+    unzip -qo /opt/valheimplus/UnixServer.zip -x changelog.txt run_bepinex.sh -d /opt/valheimplus
+    if [ -f /opt/valheim/BepInEx/config/valheim_plus.cfg ]; then
+        cp -a /opt/valheim/BepInEx/config/valheim_plus.cfg /opt/valheimplus/BepInEx/config/valheim_plus.cfg
+    fi
+    chmod 666 /opt/valheimplus/BepInEx/config/valheim_plus.cfg
+    rsync -a --itemize-changes --exclude UnixServer.zip /opt/valheimplus/ /opt/valheim
+}
+
+update_valheim_plus() {
+    remote_md5="$(curl -sL https://github.com/nxPublic/ValheimPlus/releases/latest/download/UnixServer.zip | md5sum | cut -d ' ' -f 1)"
+    local_md5="$(md5sum /opt/valheimplus/UnixServer.zip | cut -d ' ' -f 1)"
+    if [ "$remote_md5" != "$local_md5" ]; then
+        install_valheim_plus
+    else
+        echo "Valheim Plus is already the latest version"
+    fi
+}
+
+install_valheim() {
+    ./steamcmd.sh +login anonymous +force_install_dir /opt/valheim +app_update 896660 $1 $STEAMCMD_ARGS +quit
+    if [ $VALHEIM_PLUS_ENABLED = true ]; then
+        install_valheim_plus
+    fi
+    echo "Valheim Server was updated - restarting"
+    supervisorctl restart valheim-server
+}
+
+update_valheim() {
+    if [ ! -d /opt/valheim/valheim_server_Data ] || [ $FORCE_VALIDATE = true ]; then
+        echo "Installing Valheim Server"
+        args=""
+        if [ $FORCE_VALIDATE = true ]; then
+            args=-validate
+            FORCE_VALIDATE=false
+        fi
+        install_valheim $args
     else
-        echo "Valheim Server is already the latest version"
-        if [ $just_started = true ]; then
-            supervisorctl start valheim-server
+        local_version=$(sed -n '29p' /opt/valheim/steamapps/appmanifest_896660.acf | xargs | cut -d' ' -f2-)
```

Comment:
Seems rather fragile to go by line number. Can we parse the value in a more reliable way?

Reply:
How about doing this line like this:

```
awk '/"896661"/,/"manifest"/{if($1=="\"manifest\"")print $2;}' /opt/valheim/steamapps/appmanifest_896660.acf | xargs
```

So basically we use `awk` to extract everything between `"896661"` and `"manifest"`, then we check if the first column matches `"manifest"` and extract the second column. Finally we pass it to `xargs` to strip extraneous characters and return just the raw ID